packages/block-editor/src/components/observe-typing/index.js

```diff
@@ -53,58 +60,60 @@ function isKeyDownEligibleForStartTyping( event ) {
  * element.
  */
 export function useMouseMoveTypingReset() {
-	const isTyping = useSelect(
-		( select ) => select( blockEditorStore ).isTyping(),
-		[]
-	);
-	const { stopTyping } = useDispatch( blockEditorStore );
+	const { isTyping } = useSelect( blockEditorStore );
+	const { stopTyping, startTyping } = useDispatch( blockEditorStore );
 
-	return useRefEffect(
-		( node ) => {
-			if ( ! isTyping ) {
-				return;
-			}
+	return useRefEffect( ( node ) => {
+		const { ownerDocument } = node;
+		let lastClientX;
+		let lastClientY;
+		let timeoutSinceLastMouseMove;
 
-			const { ownerDocument } = node;
-			let lastClientX;
-			let lastClientY;
+		/**
+		 * On mouse move, unset typing flag if user has moved cursor.
+		 *
+		 * @param {MouseEvent} event Mousemove event.
+		 */
+		function stopTypingOnMouseMove( event ) {
+			const { clientX, clientY } = event;
 
-			/**
-			 * On mouse move, unset typing flag if user has moved cursor.
-			 *
-			 * @param {MouseEvent} event Mousemove event.
-			 */
-			function stopTypingOnMouseMove( event ) {
-				const { clientX, clientY } = event;
+			// We need to check that the mouse really moved because Safari
+			// triggers mousemove events when shift or ctrl are pressed.
+			const didMove =
+				lastClientX &&
+				lastClientY &&
+				( lastClientX !== clientX || lastClientY !== clientY );
 
-				// We need to check that the mouse really moved because Safari
-				// triggers mousemove events when shift or ctrl are pressed.
-				if (
-					lastClientX &&
-					lastClientY &&
-					( lastClientX !== clientX || lastClientY !== clientY )
-				) {
-					stopTyping();
-				}
+			if ( didMove && isTyping() ) {
+				stopTyping();
+			}
 
-				lastClientX = clientX;
-				lastClientY = clientY;
+			if ( didMove ) {
+				clearTimeout( timeoutSinceLastMouseMove );
+				timeoutSinceLastMouseMove = setTimeout( () => {
+					// Only hide the toolbar if the active element is inside the container
+					// For instance if the focus is on a popover of the toolbar,
+					// we shouldn't hide the toolbar.
+					if ( node.contains( ownerDocument.activeElement ) ) {
+						startTyping();
+					}
+				}, TOOLBAR_INACTIVITY_TIMEOUT );
```

Comment:
I think we may want to consider keeping the toolbar visible if the mouse is positioned over the toolbar. (Note that the toolbar may not be focused at this point).

https://user-images.githubusercontent.com/1270189/152062852-696e682f-f909-4d97-a032-3d459dcbe5db.mp4

Reply:
I thought about this but holistically it's not that simple:

 1- we want to keep it visible when the mouse is over the toolbar
 2- when the mouse is over a dropdown opened from the toolbar
 3- when the mouse is over a modal opened from the toolbar

In the two final cases, the popover and dropdown are not rendered "inside" the toolbar in the DOM, meaning finding a way to check for all the desired use-cases is hard.

That said, for the popover or modal to stay visible, focus need to be there (inside them) and this is covered by the check we already have in place here (focus inside the canvas or avoid hiding the toolbar) meaning, we can in theory only cover the first case and we'd be covering all use-cases.

That said, I was wondering that if someone is using the mouse to try to click the toolbar, they're probably moving the mouse, and if they stop, they won't stop long, the only use-case where they'd stop for a second is to check a tooltip, so maybe if the timeout is enough (5 seconds) to cover that, it should be enough and we don't have to care about the position of the mouse?